### PR TITLE
Feature/issue #26 user login

### PR DESCRIPTION
**Pull request by tomy-pg**

### CODE REVIEW

File: src/api/algorithms/index.ts [hunk 1]
```diff
@@ -1,17 +1,18 @@
 // tslint:disable:no-expression-statement
 import { Algorithms } from '../../application/types';
-import axios, { AxiosResponse, AxiosError } from 'axios';
+import { AxiosResponse, AxiosError } from 'axios';
```
Comment (rasmus-storjohann-PG) [inline src/api/algorithms/index.ts]: There's some food for thought here https://medium.com/frontend-digest/axios-vs-fetch-which-to-use-in-2019-6678c083c5c

File: src/application/build_types.ts [hunk 2]
```diff
@@ -79,10 +79,22 @@ export const buildInvalidServicesType = (errorMessage: string) => (
...
-);
\ No newline at end of file
+);
+
+export const buildEmptyUserType = () => (
+    helpers.makeEmptyType(constants.USER_EMPTY)
+);
+
+export const buildValidUserType = () => (
+    helpers.makeEmptyType(constants.USER_VALID)
```
Comment (rasmus-storjohann-PG) [inline src/application/build_types.ts]: It would make more sense that this type contains the token and the caller of `requestLogin()` calls `Cookies.set('token', response.data.token);`.

File: src/application/types.ts [hunk 3]
```diff
@@ -189,4 +189,19 @@ export interface ValidatedLocationJSON {
 export interface ValidatedServiceAtLocationJSON {
     readonly service: ValidatedServiceJSON;
     readonly location: ValidatedLocationJSON;
-}
\ No newline at end of file
+}
+
+export interface EmptyUser {
+    readonly type: 'USER:EMPTY';
+}
```
Comment (rasmus-storjohann-PG) [inline src/application/types.ts]: This type should be connected with `buildEmptyUserType()` I think.

File: src/api/login/index.ts [hunk 4]
```diff
@@ -3,18 +3,21 @@ import axios, { AxiosResponse, AxiosError } from 'axios';
...
     return await axios.post(url, {
         username: loginData.username,
         password: loginData.password,
     })
-    .then((response: AxiosResponse): ValidUser => {
-        Cookies.set('token', response.data.token);
-        return buildValidUserType();
+    .then((response: AxiosResponse): User => {
+        return validateLoginResponse(response);
     })
```
Comment (rasmus-storjohann-PG) [inline src/api/login/index.ts]: This is getting there...

File: src/api/login/index.ts [hunk 5]
```diff
@@ -25,4 +28,22 @@ export const buildLoginUrl = (): string => {
...
+    }
+    const validator = validateIncomingToken(response.data);
+    if (isValidationError(validator)) {
+        const errorMessage = 'Error: locations response data failed schema validation';
+        return buildInvalidUserType(errorMessage);
+    }
+    if (R.isEmpty(response.data)) {
+        return buildEmptyUserType();
+    }
+    Cookies.set('token', response.data.token);
```
Comment (rasmus-storjohann-PG) [inline src/api/login/index.ts]: Hmmmm, validator with a bit of a side effect going on here...

File: src/application/index.tsx [hunk 6]
```diff
@@ -1,20 +1,37 @@
...
         );
-};
\ No newline at end of file
+};
+export interface PrivateRouteProps {
+    readonly Component: React.ComponentType<RouterProps>;
+    readonly user: User;
+}
+
+const PrivateRoute = (props: PrivateRouteProps): JSX.Element => {
```
Comment (rasmus-storjohann-PG) [inline src/application/index.tsx]: Not familiar with the concept of a private router

File: src/application/index.tsx [hunk 7]
```diff
@@ -1,20 +1,37 @@
...
                     <Login {...props} user={user} setUser={setUser}/>} />
+                <PrivateRoute user={user} Component={ServiceRelevancyScore}/>
             </Switch>
         </Router>
         );
-};
\ No newline at end of file
+};
+export interface PrivateRouteProps {
+    readonly Component: React.ComponentType<RouterProps>;
```
Comment (rasmus-storjohann-PG) [inline src/application/index.tsx]: Don't capitalize `Component`, that is just used for class names and component names, not properties that happen to be components.

File: src/application/index.tsx [hunk 8]
```diff
@@ -20,18 +20,25 @@ export const Application = (): JSX.Element => {
...
+           renderComponentOrRedirect(props, renderProps)
         }
       />
     );
-  };
\ No newline at end of file
+  };
+
+const renderComponentOrRedirect = (props: PrivateRouteProps, renderProps: RouteChildrenProps): JSX.Element => {
+    const Component = props.Component;
```
Comment (rasmus-storjohann-PG) [inline src/application/index.tsx]: `component` here too, not `Component`

File: src/application/index.tsx [hunk 9]
```diff
@@ -2,7 +2,7 @@
 import React, { useState } from 'react';
 import { BrowserRouter as Router, Switch, Route, Redirect } from 'react-router-dom';
 import { ServiceRelevancyScore } from '../components/pages/service_relevancy_score';
-import { Login } from '../components/pages/login_page';
+import { Login } from '../components/pages/login/login';
```
Comment (rasmus-storjohann-PG) [inline src/application/index.tsx]: This path is repetitive, I'd prefer `import { Login } from '../components/pages/login';` which can be accomplished by moving the tsx file up one level or renaming it to index.ts

File: src/components/pages/login/login.tsx [hunk 10]
```diff
@@ -1,24 +1,24 @@
...
+import { SetUser, User } from '../../../application/types';
+import * as constants from '../../../application/constants';
 
 export interface LoginData {
     readonly username: string;
     readonly password: string;
 }
 
-export interface LoginProps {
+export interface UserProps {
```
Comment (rasmus-storjohann-PG) [inline src/components/pages/login/login.tsx]: The old way was right, the component is Login, so the props should be LoginProps. Consistency, least surprise.

File: src/components/pages/login/login.tsx [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+};
+
+const renderUserStatusMessage = (user: User): JSX.Element => {
+  switch (user.type) {
+    case constants.USER_VALID:
+      return <h2>Login Success!</h2>;
+    case constants.USER_INVALID:
+    return <h2>Username/Password invalid. Please try again. </h2>;
+    default:
+      return <div></div>;
```
Comment (rasmus-storjohann-PG) [inline src/components/pages/login/login.tsx]: `return <div/>;`

File: src/application/index.tsx [hunk 12]
```diff
@@ -15,18 +15,23 @@ export const Application = (): JSX.Element => {
...
 };
 
-export interface PrivateRouteProps {
-    readonly Component: React.ComponentType<RouterProps>;
+export interface UserProps {
     readonly user: User;
+    readonly setUser: SetUser;
+}
+export interface PrivateRouteProps {
+    readonly Component: React.ComponentType<RouterProps&UserProps>;
```
Comment (rasmus-storjohann-PG) [inline src/application/index.tsx]: space around &

File: src/application/index.tsx [hunk 13]
```diff
@@ -15,18 +15,23 @@ export const Application = (): JSX.Element => {
...
 }
 
-const PrivateRoute = (props: PrivateRouteProps): JSX.Element => {
+type Props = UserProps & PrivateRouteProps;
```
Comment (rasmus-storjohann-PG) [inline src/application/index.tsx]: This repeats from `React.ComponentType<RouterProps&UserProps>` above, DRY it up

File: src/components/logout/logout_button.tsx [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+import { UserProps } from '../../application';
+import { buildEmptyUserType } from '../../application/build_types';
+
+export const LogoutButton = (props: UserProps): JSX.Element => (
+    <button onClick={(): void => logoutUser(props)}>Logout</button>
+);
+
+export const logoutUser = (props: UserProps): void => (
+    props.setUser(buildEmptyUserType())
+);
```
Comment (rasmus-storjohann-PG) [inline src/components/logout/logout_button.tsx]: No need to export this.

File: src/components/page_switcher/private_route.tsx [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    );
+  };
+
+const renderComponentOrRedirect = (props: Props, renderProps: RouteChildrenProps): JSX.Element => {
+    const Component = props.Component;
+    if (props.user.type !== constants.USER_VALID) {
+        return <Redirect to='/login'/>;
+    }
+    return <Component {...props} {...renderProps}/>;
+};
```
Comment (rasmus-storjohann-PG) [inline src/components/page_switcher/private_route.tsx]: Keep variables' scope as small as possible, move `const Component = props.Component;` below the early return.

File: src/components/page_switcher/page_switcher.tsx [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+import { ServiceRelevancyScore } from '../pages/service_relevancy_score';
+import { Login } from '../pages/login/login';
+import { PrivateRoute } from './private_route';
+
+export const PageSwitcher = (props: UserProps): JSX.Element => (
+    <Switch>
+        <Route exact path={'/login'} render={(renderProps: RouteChildrenProps): JSX.Element =>
+            <Login {...renderProps} user={props.user} setUser={props.setUser}/>} />
+        <PrivateRoute user={props.user} setUser={props.setUser} Component={ServiceRelevancyScore}/>
+    </Switch>
```
Comment (rasmus-storjohann-PG) [inline src/components/page_switcher/page_switcher.tsx]: The render function makes this really hard to follow. Pull the render function out into a local variable above the switch.

File: src/api/axios_config.ts [hunk 17]
```diff
@@ -1,7 +1,7 @@
 import axios from 'axios';
 import Cookies from 'js-cookie';
 
-export const axiosRequest = axios.create({
+export const authenticatedAxiosRequest = axios.create({
```
Comment (rasmus-storjohann-PG) [inline src/api/axios_config.ts]: `axios.create()` does not return a request, it returns something you can use to make requests.

File: src/api/services.ts [hunk 18]
```diff
@@ -15,7 +15,7 @@ export const requestServices = async (topic: ValidTopicId, location: Location, a
     return await axios.get(url)
     .then((response: AxiosResponse): Services => {
       return validateServicesResponse(response);
-  });
+  }).catch((error: AxiosError): Services => buildInvalidServicesType(error.message));
```
Comment (rasmus-storjohann-PG) [inline src/api/services.ts]: I encourage you to think more functionally, specifically to always remember that functions are first class objects in TS. In this case, that means that you don't need to make a closure to call a function. Rather than this
```ts
}).catch((error: AxiosError): Services => buildInvalidServicesType(error.message));
```
you can do this
```ts
}).catch(buildInvalidServicesType);
```
provided that you just change `buildInvalidServicesType` to be `buildInvalidServicesType(error: AxiosError)`. Nicer, no?

Comment (rasmus-storjohann-PG) [reply src/api/services.ts]: This applies widely in both the QA tool and the app. @ian-lin-PG you should pick up this habit too.

File: src/application/index.tsx (hunk 6 above)
Comment (tomy-pg) [reply src/application/index.tsx]: With the private route, I am preventing people from directly going to the Services Relevancy Score page without first being logged. 

https://tylermcginnis.com/react-router-protected-routes-authentication/

File: src/api/algorithms/index.ts (hunk 1 above)
Comment (tomy-pg) [reply src/api/algorithms/index.ts]: This was an interesting read. To be honest, my decision to use Axios was based on the needing to use less boilerplate. I did not realize it is not a very well maintained library.

File: src/application/index.tsx (hunk 7 above)
Comment (tomy-pg) [reply src/application/index.tsx]: When I use lowercase this I receive the following error:

`Property 'foo' does not exist on type 'JSX.IntrinsicElements'.`

When I capitalize `Foo` I no longer see this error.

It appears that properties that are components need to follow the custom component convention of being capitalized.

File: src/api/services.ts (hunk 18 above)
Comment (tomy-pg) [reply src/api/services.ts]: This is a lot nicer. 

However, when changing `buildInvalidServicesType` to expect the type `AxiosError` things seem to fall apart. 

Without this change, the functionality of this did not breakdown and I still received the error messages without telling the function to expect `AxiosError` as the argument type.

File: src/api/login/index.ts [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+        username: loginData.username,
+        password: loginData.password,
+    })
+    .then((response: AxiosResponse): void => console.log(response.data))
+    .catch((error: AxiosError): void =>  console.log(error.message));
+};
+
+export const buildLoginUrl = (): string => {
+    const path = 'authenticate/';
+    const baseUrl = process.env.REACT_APP_API_URL || 'http://127.0.0.1:8000/';
```
[the remark this answers is not in the document]
Comment (tomy-pg) [reply src/api/login/index.ts]: Without having this fallback value, baseURL is viewed as being either a string or undefined which breaks the up on build.
